migrations/multiple_contract_migration.go

```diff
@@ -0,0 +1,230 @@
+	}
+
+	if len(program.Declarations) == 1 {
+		// If there is one declaration move it to the new key
+		m.logger.Debug().
+			Str("address", address.Hex()).
+			Msg("Single contract or interface at address moved to new key")
+		p.Key = addNameToKey(p.Key, program.Declarations[0].DeclarationIdentifier().Identifier)
+		return []ledger.Payload{p}, nil
+	}
+
+	// We have two declarations. Due to the current rules one of them is an interface and one is a contract.
+	// the contract will need an import to the interface.
+	m.logger.Info().
+		Str("address", address.Hex()).
+		Msg("Two declarations on an address, splitting into two parts")
+
+	_, oneIsInterface := program.Declarations[0].(*ast.InterfaceDeclaration)
+	_, twoIsInterface := program.Declarations[1].(*ast.InterfaceDeclaration)
+	if oneIsInterface == twoIsInterface {
+		// declarations of same type! should not happen
+		m.logger.Error().
+			Str("address", address.Hex()).
+			Str("code", code).
+			Msg("Two declarations of the same type at address")
+		return []ledger.Payload{}, fmt.Errorf("wwo declarations of the same type at address %s", address.Hex())
+	}
+
+	var interfaceDeclaration ast.Declaration
+	var contractDeclaration ast.Declaration
+	if oneIsInterface {
+		interfaceDeclaration = program.Declarations[0]
+		contractDeclaration = program.Declarations[1]
+	} else {
+		interfaceDeclaration = program.Declarations[1]
+		contractDeclaration = program.Declarations[0]
+	}
+
+	interfaceStart := interfaceDeclaration.StartPosition().Offset
+	interfaceEnd := interfaceDeclaration.EndPosition().Offset
+	contractStart := contractDeclaration.StartPosition().Offset
+	contractEnd := contractDeclaration.EndPosition().Offset
+
+	var contractCode string
+	var interfaceCode string
+
+	if contractStart < interfaceStart {
+		split := contractEnd + 1
+		contractCode = code[:split]
+		interfaceCode = code[split:]
+	} else {
+		split := interfaceEnd + 1
+		contractCode = code[split:]
+		interfaceCode = code[:split]
+	}
+
+	// add import to contract code
+	contractCode = fmt.Sprintf("import %s from %s\n%s", interfaceDeclaration.DeclarationIdentifier().Identifier, address.HexWithPrefix(), contractCode)
+
+	interfaceKey := addNameToKey(p.Key, interfaceDeclaration.DeclarationIdentifier().Identifier)
+	contractKey := addNameToKey(p.Key, contractDeclaration.DeclarationIdentifier().Identifier)
+	return []ledger.Payload{
+		{
+			Key:   interfaceKey,
+			Value: []byte(interfaceCode),
+		}, {
+			Key:   contractKey,
+			Value: []byte(contractCode),
+		},
+	}, nil
+}
+
+// if it is a address
+func needsMigration(id flow.RegisterID) bool {
+	if id.Key != "code" {
+		return false
+	}
+	return len([]byte(id.Owner)) == flow.AddressLength
+}
+
+func addNameToKey(key ledger.Key, name string) ledger.Key {
+	newKey := key
+	newKey.KeyParts = make([]ledger.KeyPart, 3)
+	copy(newKey.KeyParts, key.KeyParts)
+	newKeyString := fmt.Sprintf("code.%s", name)
```

Comment:
the long term plan was "code" be a key part type and we only use the name, so the ledger can optimize the code storage but I'm fine with this migration for now.